### PR TITLE
ios14 migration (rework)

### PR DESCRIPTION
**Pull request by wutschel**
Basic rework to migrate Kodi Remote App to iOS 14, building with XCode 12.2/.3. Summary of changes:

- Updated target properties to fulfill iOS14 requirements (local network access, registering bonjour services)
- Migrate to NSJSONSerialization (fixes main problems with crashes in several uses cases)
- Several clean ups (fixed runtime and compile time warnings, replaced deprecated function calls)
- Fixes / fine tuning for several views and items for iPad and iPhone
- Reworked as outcome of https://github.com/xbmc/Official-Kodi-Remote-iOS/pull/101

ToDo:

- Search, Collection View and Sorting-By does not work (rework started on another branch)
- UIWebView must be replaced to allow upload to AppStore (only can offer a full removal for now)

### CODE REVIEW

File: XBMC Remote/VolumeSliderView.m [hunk 1]
```diff
@@ -52,14 +52,14 @@ - (id)initWithFrame:(CGRect)frame{
...
-            [plusButton setBackgroundImage:[UIImage imageNamed:@"button_volume_plus"] forState:UIControlStateNormal];
             [plusButton setBackgroundImage:[UIImage imageNamed:@""] forState:UIControlStateHighlighted];
+            [plusButton setBackgroundImage:[UIImage imageNamed:@"button_volume_plus.png"] forState:UIControlStateNormal];
             [plusButton setTitle:@"" forState:UIControlStateNormal];
             [plusButton setTitle:@"" forState:UIControlStateHighlighted];
             [plusButton setShowsTouchWhenHighlighted:YES];
 
-            [minusButton setBackgroundImage:[UIImage imageNamed:@"button_volume_minus"] forState:UIControlStateNormal];
             [minusButton setBackgroundImage:[UIImage imageNamed:@""] forState:UIControlStateHighlighted];
+            [minusButton setBackgroundImage:[UIImage imageNamed:@"button_volume_minus.png"] forState:UIControlStateNormal];
```
Comment (kambala-decapitator) [inline XBMC Remote/VolumeSliderView.m]: could you keep the order of lines in both hunks?

**Review by kambala-decapitator — CHANGES_REQUESTED** [review]

File: XBMC Remote/iPad/StackScrollViewController.m [hunk 2]
```diff
@@ -272,9 +272,9 @@ -(void)arrangeVerticalBar {
 }
 
 -(void)offView{
-    
-    int orientation= [[UIApplication sharedApplication] statusBarOrientation];
-    int posX = (orientation==1 || orientation==2) ? 468 : 724;
+
```
Comment (kambala-decapitator) [inline XBMC Remote/iPad/StackScrollViewController.m:275]: adding whitespace to an empty line isn't so good :)

**Comment by kambala-decapitator** [comment]
Commit message of https://github.com/xbmc/Official-Kodi-Remote-iOS/pull/102/commits/088fbccd563aaa34da7cc92b13b1532858677d8b is slightly incorrect now: there's no more `#define` :)

File: XBMC Remote/ShowInfoViewController.m [hunk 3]
```diff
@@ -1786,18 +1795,22 @@ - (void)showBackground:(id)sender {
...
+    if (arrow_continue_down.hidden!=YES && at_bottom==YES){
+        arrow_continue_down.hidden=YES;
         [UIView beginAnimations:nil context:nil];
         [UIView setAnimationCurve:UIViewAnimationCurveEaseOut];
         [UIView setAnimationDuration:1];
-        arrow_continue_down.alpha=0;
         [UIView commitAnimations];
     }
-    else if (arrow_continue_down.alpha==0 && theScrollView.contentOffset.y<40){
+    else if (arrow_continue_down.hidden==YES && at_bottom!=YES){
```
Comment (kambala-decapitator) [inline XBMC Remote/ShowInfoViewController.m:1809]:
```suggestion
    else if (arrow_continue_down.hidden && !at_bottom){
```

File: XBMC Remote/ShowInfoViewController.m [hunk 4]
```diff
@@ -1786,18 +1795,22 @@ - (void)showBackground:(id)sender {
 }
 
 - (void) scrollViewDidScroll: (UIScrollView *) theScrollView{
-    if (arrow_continue_down.alpha && theScrollView.contentOffset.y>40){
+    int height_content = theScrollView.contentSize.height;
+    int height_bounds = theScrollView.bounds.size.height;
+    int scrolled = theScrollView.contentOffset.y;
+    bool at_bottom = scrolled >= height_content-height_bounds;
+    if (arrow_continue_down.hidden!=YES && at_bottom==YES){
```
Comment (kambala-decapitator) [inline XBMC Remote/ShowInfoViewController.m:1802]:
```suggestion
    if (!arrow_continue_down.hidden && at_bottom){
```

File: XBMC Remote/ShowInfoViewController.m [hunk 5]
```diff
@@ -611,7 +611,10 @@ -(void)recordChannel {
 }
 
 -(IBAction)scrollDown:(id)sender{
-    CGPoint bottomOffset = CGPointMake(0, scrollView.contentSize.height - scrollView.bounds.size.height);
+    int height_content = scrollView.contentSize.height;
+    int height_bounds = scrollView.bounds.size.height;
+    int bottom_scroll = (height_content - height_bounds) > 0 ? height_content - height_bounds : 0;
```
Comment (kambala-decapitator) [inline XBMC Remote/ShowInfoViewController.m:616]:
```suggestion
    int bottom_scroll = MAX(height_content - height_bounds, 0);
```

**Review by kambala-decapitator — CHANGES_REQUESTED** [review]

**Comment by kambala-decapitator** [comment]
apart from minor stuff outlined above, looks pretty well!

File: XBMC Remote/VolumeSliderView.m (hunk 1 above)
Comment (wutschel) [reply XBMC Remote/VolumeSliderView.m]: Argh, I was overseeing this. Absolutely not sure why I did this. Interesting is that I am fixing this back to the original order with the next commit. Shall I fix both commits through squashing? Or do we keep it as the order is back to normal after 2nd commit?

Comment (kambala-decapitator) [reply XBMC Remote/VolumeSliderView.m]: I'd prefer to have the right order in the first commit, but since next commit fixes it, it's also fine.

**Comment by wutschel** [comment]
Here we go. Even fixed the ordering issues in the first commits. :)

**Review by kambala-decapitator — APPROVED** [review]
nice!

**Comment by kambala-decapitator** [comment]
@fuzzard got anything to add?

**Comment by fuzzard** [comment]
I'm not able to do much ATM. At a glance, looks fine. Happy to merge and get the stuff in. Runtime test the other day has it at least functional on 14. Not able to do store release yet, but this is definitely a great start. Thanks @wutschel

**Comment by wutschel** [comment]
@kambala-decapitator and @fuzzard: Thanks for the support!

Few thoughts on releasing: 
1. The App will most likely not be accepted as UIWebView is still included. I can provide a PR which removes this completely. Instead of showing content (e.g. Wikipedia search or embedded movie trailers) it will call the system browser. For the movie trailer related changes it would be good to have someone testing this. Alternatively we migrate to the new WKWebView. I cannot judge how much work this would be.
2. The App in its current state still faces problems with the Search bar and buttons placed there. As kambala-decapitator found out, the Search problem starts with iOS 13. As a first quick hack we can disable the Search bar for iOS 13+ for now. kambala-decapitator already proposed a workaround to achieve this. This workaround only needs some beautification to avoid an empty black bar on top of the screen (I have an idea what to do).
3. Migrate from UISeachDisplayController to UISearchController. This is quite some work, and I already have a working version which allows search for iPhone and also re-enables the SearchBar buttons (to switch between list/collection view and to change the sorting). But this has some nasty hacks and I need support to do this properly. I expect that we will see lots of code which only works for iOS 13+. Do we possibly only update the App for iOS 13+? This way we do not touch the setups which still work and we can further clean up the App code.

**Comment by fuzzard** [comment]
Wkwebview migration doesnt look like it would be horrible, but if we just need a working remote, a PR for removal of uiwebview until a later time wkwebview can be implemented may be best.

I think bumping the mintargetversion probably isn't desired. I'm usually the first to say drop old versions in Kodi core, but for a remote, I think trying to keep backwards compat is more ideal. As painful as that is, I think we are best to just API/version check codepaths.

For the search, pr what you have, let's have a look, and see what we can do to help.

**Comment by fuzzard** [comment]
I'll also throw this offer to you, if you are interested we can give you an invite to a Kodi team Slack channel. May make dialog a bit quicker/easier, but up to you. Just let us know

**Comment by kambala-decapitator** [comment]
@wutschel 

1. UIWebView must be replaced definitely. Would it work if you use `SFSafariViewController` instead of external browser? UX would be better IMO.
2. If we need to deliver AppStore update ASAP (not aware of any functionality issues with the current live app as I don't use it), then this is fine. Otherwise let's make proper implementation with `UISearchController`.
3. As long as supporting iOS 9-12 is trivial, we should keep it. For example, #95 definitely deserves to be available to old devices.

**Comment by wutschel** [comment]
@kambala-decapitator and @fuzzard

1. PR https://github.com/xbmc/Official-Kodi-Remote-iOS/pull/103 provides the removal of `UIWebView`. We can further discuss in the new PR.
2. PR https://github.com/xbmc/Official-Kodi-Remote-iOS/pull/104 is the current state of affairs for the `UISearchController`. You will see there is a lot left for rework...
3. Ok, then let´s do our best to keep at least iOS 9+ supported. Looking at the warnings it seems there are less cases than I initially remembered (only several occasions of`showsSearchResultsController` and `searchTextField`). This looks manageable.

On the AppStore update: The current App is not really usable since iOS 14. Main problem are crashes which are fixed with the JSON migration in this PR. These crashes occur for a lot of standard use cases and functions (like updating the library). But, we cannot update the App quickly without breaking Search. Or is there a way to build with an older SDK version and release it?

**Comment by kambala-decapitator** [comment]
thanks. We're waiting for the app owner to transfer the app to Kodi's developer account, then we can manage App Store releases.

> Or is there a way to build with an older SDK version and release it?

currently the earliest allowed toolchain is Xcode 11 / iOS 13 SDK. Does search work on iOS 13+ when built with it? (although I doubt)

**Comment by kambala-decapitator** [comment]
let's continue discussion in respective PRs

**Pull request merged**